src/OpalCompiler-Core/OCPseudoVariableScope.class.st

```diff
@@ -0,0 +1,20 @@
+"
+I model a scope that resolves pharo pseudo-variables such as: `self`, `super` and `thisContext`.
+
+I should by default put at the beginning of the chain when resolving names so these names are resolved with more priority than others.
+"
+Class {
+	#name : 'OCPseudoVariableScope',
+	#superclass : 'OCAbstractScope',
+	#category : 'OpalCompiler-Core-Semantics',
+	#package : 'OpalCompiler-Core',
+	#tag : 'Semantics'
+}
+
+{ #category : 'lookup' }
+OCPseudoVariableScope >> lookupVar: name [
+
+	^ PseudoVariable lookupDictionary
```

Comment:
A problem here is that you create new PseudoVariables for evert method. This is problematic, as we now have one instance for every definition of a variable, and we use that e.g to for #usingMethods (if not overridden to use the bytecode for speed) and for AST related queries (#readNodes, for example).

Of course this is a question of definition, but this was one of the nice aspects of how it is now.

Reply:
e.g. you can do now 
```
(Smalltalk globals lookupVar: #thisProcess) readNodes
```

(it is slow as it has to create the AST for every method)

Reply:
Another nice aspect of the current way  (and maybe one of the main reasons why it is done like that): you can lookup the one instance of a pseudo variable reflectively outside of the compiler. If the PseudoVariableScope is created just for compilation, this is not possible. 

```

var :=  (Smalltalk globals lookupVar: #self).

"now I can use it reflectively, in the doit it is nil of course"

var readInContext: thisContext.

"like every var it knows it's scope, if I look my name up I get myself"

(var scope lookupVar: var name) == var "true"
```

Reply:
But making the scope of PseudoVariables explicit is nice... one idea: could it not be the parent or immediate child scope of the global scope (depending if we want to allow them to be shadowed from the global or not).

In the current setup, the scopes of Variables that are always allocated (Slots, Class Vars, Globals) are explicit, always allocated scopes. They are not subclass of OCAbstractScope, but they do implement the API and *are* the scopes of the Variables (they form the scope hierarchy directly above the method scope).

```
var := Smalltalk globals lookupVar: #Array.
var class "GlobalVariable"
var scope.  "a SystemDictionary(lots of globals)"

var := SmalltalkImage classVariableNamed: #CompilerClass.
var class  "ClassVariable"
"the scope of a Class Variable is the class"
var scope  "SmalltalkImage" 
```

This for sure we do not want to loose if we get rid of the global scope. E.g. I would assume that the Variables that are defined by a "Module" would have that module as their scope.

In a way, the PseudoVariables are in a not-explicit child scope of the global scope right now, but we could make it explicit and have it as a child or parent scope now of the global scope.  

For"Module" scopes I guess one could have one instance of the PseudoVariableScope per module.

Reply:
I do not really get why the namespace is responsible for the scope of self. 
I do not understand why self is not looked up first. Because we should not be able to define an instance variable self or a classvariable. Could we make sure that we do not introduce more ways to break the code. So I do not get why having a scope for self by method is a problem.
When we compile a method then the method has a scope and this scope contains self 
I do not understand why something simple like that is not simple.

Reply:
We should discuss.